libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/arrow/api/MapboxRouteArrowApi.kt

```diff
@@ -61,101 +70,194 @@ import com.mapbox.turf.TurfMeasurement
+                is Expected.Failure -> {
+                    addArrowResult
+                }
+                is Expected.Success -> {
+                    Expected.Success(
+                        UpdateManeuverArrowValue(
+                            visibilityChanges,
+                            addArrowResult.value
+                                .arrowShaftFeatureCollection
+                                .features()
+                                ?.firstOrNull(),
+                            addArrowResult.value
+                                .arrowHeadFeatureCollection
+                                .features()
+                                ?.firstOrNull()
+                        )
+                    )
+                }
+            }
+        }
+
+    /**
+     * Adds an arrow to the map made up of the points submitted. An arrow is made up of at least
+     * two points. The direction of the arrow head is determined by calculating the bearing
+     * between the last two points submitted. Each call will add a new arrow.
+     *
+     * @param points the points that make up the arrow to be drawn
+     *
+     * @return a Expected<ArrowAddedValue, RouteArrowUpdateError>
+     */
+    fun addArrow(points: List<Point>): Expected<ArrowAddedValue, RouteArrowUpdateError> {
+        if (points.size < RouteConstants.TWO_POINTS) {
+            return Expected.Failure(
+                RouteArrowUpdateError.InvalidPointError(
+                    "An arrow must have at least 2 points.",
+                    null
+                )
             )
-        val invalidCurrentStepPoints = routeProgress.currentLegProgress == null ||
-            routeProgress.currentLegProgress!!.currentStepProgress == null ||
-            routeProgress.currentLegProgress!!.currentStepProgress!!.stepPoints == null ||
-            routeProgress.currentLegProgress!!.currentStepProgress!!.stepPoints!!.size <
-            RouteConstants.TWO_POINTS
-
-        val visibilityChanges = if (invalidUpcomingStepPoints || invalidCurrentStepPoints) {
-            getHideArrowModifications()
-        } else {
-            getShowArrowModifications()
         }
 
-        val newManeuverPoints = if (!invalidUpcomingStepPoints && !invalidCurrentStepPoints) {
-            RouteArrowUtils.obtainArrowPointsFrom(routeProgress)
+        return if (arrows.flatten().intersect(points).isEmpty()) {
+            arrows.add(points)
+            return Expected.Success(
+                ArrowAddedValue(
+                    getShaftFeatureCollection(),
+                    getArrowHeadFeatureCollection()
+                )
+            )
         } else {
-            listOf()
+            Expected.Failure(
+                RouteArrowUpdateError.AlreadyPresentError(
+                    "An arrow with one or more of those points already exists",
+                    null
+                )
+            )
+        }
```

Comment:
Should we allow this and let the developer to handle it? I mean, it's up to the developer what they want to add right? Even if previously added 🤔

Reply:
I don't see any value to the developer in allowing this and instead it seems like a potential source for bugs.  Let's say there is a bug in the developer's app that adds the the same arrow twice. Then when they call remove, the arrow would still appear because there are two of them.  This leads to a bug reported to us. A bug that would be very difficult to remedy since the bug is in the developer's app.  

If we get a request to allow overlapping arrows then we'll make the adjustment but I think by supporting this in the beginning it's unnecessary complexity.

Reply:
What happens if they want to display 2 arrows for a right and left intersection? It seems that currently that's not an option.

Reply:
Hey @mapbox/navigation-ios are you implementing this? If so, how are you doing it? Any thoughts on supporting overlapping arrows?

cc @avi-c @d-prukop

Reply:
On iOS, we aren’t currently displaying multiple arrows along the route line. Instead, the arrow’s shaft extends a fixed distance along the route line, traversing whatever bends and maneuvers may lie along the way. It probably would be clearer to show multiple arrows as this PR would: mapbox/mapbox-navigation-ios#2869. I think it’s perfectly reasonable to allow earlier arrows to overlap later arrows, as long as an earlier arrowhead is placed before a later arrowhead. Apple Maps does this, at least on the preview map.

Reply:
After refactor overlapping arrows are possible.